src/wpublish/QueriesAndChunks.cc

```diff
@@ -601,21 +674,21 @@ vector<wbase::Task::Ptr> QueriesAndChunks::removeQueryFrom(QueryId const& qId,
     vector<wbase::Task::Ptr> removedList;  // Return value;
 
     // Find the user query.
-    unique_lock<mutex> lock(_queryStatsMtx);
-    auto query = _queryStats.find(qId);
-    if (query == _queryStats.end()) {
+    auto query = getStats(qId);
+    if (query == nullptr) {
         LOGS(_log, LOG_LVL_DEBUG, "was not found by removeQueryFrom");
         return removedList;
     }
-    lock.unlock();
 
     // Remove Tasks from their scheduler put them on 'removedList', but only if their Scheduler is the same
     // as 'sched' or if sched == nullptr.
     vector<wbase::Task::Ptr> taskList;
     {
-        lock_guard<mutex> taskLock(query->second->_qStatsMtx);
-        taskList = query->second->_tasks;
+        lock_guard<mutex> taskLock(query->_qStatsMtx);
+        taskList = query->_tasks;
```

Comment:
This coding pattern is dangerous since the calling context is accessing and using private data members of another object. Why not to redesign that `query` object by adding the following method:
```
class Query {
public:
    <task-list> getTasks() const {
        lock_guard<mutex> taskLock(_qStatsMtx);
        return_tasks;
   }
```
After that, you may simply call that method:
```
     taskList = query->getTasks();
```

Reply:
Agreed. Originally QueryStatistics was a small struct, and this is leftover from that time. QueryStatistics is now pretty big and should no longer be a friend to QueriesAndChunks.